File: Tween/ImageDictionary.vb
Imports System.Drawing
Imports System.IO
Imports System.Threading

Public Class ImageDictionary
    Implements IDictionary(Of String, Image), IDisposable

    Private ReadOnly lockObject As New Object()

    Private memoryCacheCount As Integer
    Private innerDictionary As Dictionary(Of String, Image)
    Private sortedKeyList As List(Of String)    '古いもの順
    Private waitStack As Stack(Of KeyValuePair(Of String, Action(Of Image)))

    Public Sub New(ByVal memoryCacheCount As Integer)
        SyncLock Me.lockObject
            Me.innerDictionary = New Dictionary(Of String, Image)(memoryCacheCount + 1)
            Me.sortedKeyList = New List(Of String)(memoryCacheCount + 1)
            Me.memoryCacheCount = memoryCacheCount
            Me.waitStack = New Stack(Of KeyValuePair(Of String, Action(Of Image)))
        End SyncLock
    End Sub

    Public Sub Add(ByVal item As System.Collections.Generic.KeyValuePair(Of String, Image)) Implements System.Collections.Generic.ICollection(Of System.Collections.Generic.KeyValuePair(Of String, Image)).Add
        Me.Add(item.Key, item.Value)
    End Sub

    Public Sub Add(ByVal key As String, ByVal value As Image) Implements System.Collections.Generic.IDictionary(Of String, Image).Add
        SyncLock Me.lockObject
            If Me.innerDictionary.ContainsKey(key) Then Exit Sub
            Me.innerDictionary.Add(key, value)
            Me.sortedKeyList.Add(key)
            Me.waitStack.Push(New KeyValuePair(Of String, Action(Of Image))(key, Nothing))
            Me.DisposeOldImage()
        End SyncLock
    End Sub

    Public Function Remove(ByVal item As System.Collections.Generic.KeyValuePair(Of String, Image)) As Boolean Implements System.Collections.Generic.ICollection(Of System.Collections.Generic.KeyValuePair(Of String, Image)).Remove
        Return Me.Remove(item.Key)
    End Function

    Public Function Remove(ByVal key As String) As Boolean Implements System.Collections.Generic.IDictionary(Of String, Image).Remove
        SyncLock Me.lockObject
            Me.sortedKeyList.Remove(key)
            Me.innerDictionary(key).Dispose()
            Return Me.innerDictionary.Remove(key)
        End SyncLock
    End Function

    Default ReadOnly Property Item(ByVal key As String, ByVal callBack As Action(Of Image)) As Image
        Get
            SyncLock Me.lockObject
                Me.sortedKeyList.Remove(key)
                Me.sortedKeyList.Add(key)
                If Me.innerDictionary(key) IsNot Nothing Then
                    callBack(New Bitmap(Me.innerDictionary(key)))
                Else
                    'スタックに積む
                    Me.waitStack.Push(New KeyValuePair(Of String, Action(Of Image))(key, callBack))
                End If
                Me.DisposeOldImage()
            End SyncLock

            Return Nothing
        End Get
    End Property

    Default Public Property Item(ByVal key As String) As Image Implements System.Collections.Generic.IDictionary(Of String, Image).Item
        Get
            SyncLock Me.lockObject
                Me.sortedKeyList.Remove(key)
                Me.sortedKeyList.Add(key)
                Me.DisposeOldImage()
                If Me.innerDictionary(key) IsNot Nothing Then
                    Return New Bitmap(Me.innerDictionary(key))
                Else
                    Return Nothing
                End If
            End SyncLock
        End Get
        Set(ByVal value As Image)
            SyncLock Me.lockObject
                Me.sortedKeyList.Remove(key)
                Me.sortedKeyList.Add(key)
                Me.DisposeOldImage()
                Me.innerDictionary(key).Dispose()
                Me.innerDictionary(key) = value
            End SyncLock
        End Set
    End Property

    Public Sub Clear() Implements System.Collections.Generic.ICollection(Of System.Collections.Generic.KeyValuePair(Of String, Image)).Clear
        SyncLock Me.lockObject
            For Each value As Image In Me.innerDictionary.Values
                value.Dispose()
            Next

            Me.innerDictionary.Clear()
            Me.sortedKeyList.Clear()
        End SyncLock
    End Sub

    Public Function Contains(ByVal item As System.Collections.Generic.KeyValuePair(Of String, Image)) As Boolean Implements System.Collections.Generic.ICollection(Of System.Collections.Generic.KeyValuePair(Of String, Image)).Contains
        SyncLock Me.lockObject
            Return Me.innerDictionary.ContainsKey(item.Key) AndAlso Me.innerDictionary(item.Key) Is item.Value
        End SyncLock
    End Function

    Public Sub CopyTo(ByVal array() As System.Collections.Generic.KeyValuePair(Of String, Image), ByVal arrayIndex As Integer) Implements System.Collections.Generic.ICollection(Of System.Collections.Generic.KeyValuePair(Of String, Image)).CopyTo
        SyncLock Me.lockObject
            Dim index As Integer = arrayIndex
            For Each Item As KeyValuePair(Of String, Image) In Me.innerDictionary
                If array.Length - 1 < index Then
                    Exit For
                End If

                array(index) = New KeyValuePair(Of String, Image)(Item.Key, Item.Value)
                index += 1
            Next
        End SyncLock
    End Sub

    Public ReadOnly Property Count As Integer Implements System.Collections.Generic.ICollection(Of System.Collections.Generic.KeyValuePair(Of String, Image)).Count
        Get
            SyncLock Me.lockObject
                Return Me.innerDictionary.Count
            End SyncLock
        End Get
    End Property

    Public ReadOnly Property IsReadOnly As Boolean Implements System.Collections.Generic.ICollection(Of System.Collections.Generic.KeyValuePair(Of String, Image)).IsReadOnly
        Get
            Return False
        End Get
    End Property

    Public Function ContainsKey(ByVal key As String) As Boolean Implements System.Collections.Generic.IDictionary(Of String, Image).ContainsKey
        Return Me.innerDictionary.ContainsKey(key)
    End Function

    Public ReadOnly Property Keys As System.Collections.Generic.ICollection(Of String) Implements System.Collections.Generic.IDictionary(Of String, Image).Keys
        Get
            SyncLock Me.lockObject
                Return Me.innerDictionary.Keys
            End SyncLock
        End Get
    End Property

    Public Function TryGetValue(ByVal key As String, ByRef value As Image) As Boolean Implements System.Collections.Generic.IDictionary(Of String, Image).TryGetValue
        SyncLock Me.lockObject
            If Me.innerDictionary.ContainsKey(key) Then
                value = Me.innerDictionary(key)
                Return True
            Else
                Return False
            End If
        End SyncLock
    End Function

    Public ReadOnly Property Values As System.Collections.Generic.ICollection(Of Image) Implements System.Collections.Generic.IDictionary(Of String, Image).Values
        Get
            SyncLock Me.lockObject
                Dim imgList As New List(Of Image)(Me.memoryCacheCount)
                For Each value As Image In Me.innerDictionary.Values
                    imgList.Add(value)
                Next
                Return imgList
            End SyncLock
        End Get
    End Property

    Public Function GetEnumerator() As System.Collections.Generic.IEnumerator(Of System.Collections.Generic.KeyValuePair(Of String, Image)) Implements System.Collections.Generic.IEnumerable(Of System.Collections.Generic.KeyValuePair(Of String, Image)).GetEnumerator
        Throw New NotImplementedException()
    End Function

    Public Function GetEnumerator1() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
        Throw New NotImplementedException()
    End Function

    Public Sub Dispose() Implements IDisposable.Dispose
        SyncLock Me.lockObject
            For Each item As Image In Me.innerDictionary.Values
                If item IsNot Nothing Then
                    item.Dispose()
                End If
            Next
        End SyncLock
    End Sub

    Private Sub DisposeOldImage()
        If Me.sortedKeyList.Count > Me.memoryCacheCount Then
            Dim key As String = Me.sortedKeyList(Me.sortedKeyList.Count - Me.memoryCacheCount - 1)
            If Me.innerDictionary(key) IsNot Nothing Then
                'Me.innerDictionary(key).Dispose()
                Me.innerDictionary(key) = Nothing
            End If
        End If
    End Sub

    '取得一時停止
    Private _pauseGetImage As Boolean = False
    Public Property PauseGetImage As Boolean
        Get
            Return Me._pauseGetImage
        End Get
        Set(ByVal value As Boolean)
            Me._pauseGetImage = value

            Static popping As Boolean = False

            If Not Me._pauseGetImage AndAlso Not popping AndAlso Me.waitStack.Count > 0 Then
                popping = True
                '最新から処理し
                Dim imgDlProc As ThreadStart
                imgDlProc = Sub()
                                While Me.waitStack.Count > 0 AndAlso Not Me._pauseGetImage
                                    Dim req As KeyValuePair(Of String, Action(Of Image))
                                    SyncLock lockObject
                                        req = Me.waitStack.Pop
                                    End SyncLock
                                    Dim proc As New GetImageDelegate(AddressOf GetImage)
                                    proc.BeginInvoke(req, Nothing, Nothing)
                                End While
                                popping = False
                            End Sub
                imgDlProc.BeginInvoke(Nothing, Nothing)
            End If
        End Set
    End Property
    Delegate Sub GetImageDelegate(ByVal arg1 As KeyValuePair(Of String, Action(Of Image)))
    Private Sub GetImage(ByVal downloadAsyncInfo As KeyValuePair(Of String, Action(Of Image)))
        Dim callbackImage As Image = Nothing
        SyncLock lockObject
            If Me.innerDictionary(downloadAsyncInfo.Key) IsNot Nothing Then
                callbackImage = New Bitmap(Me.innerDictionary(downloadAsyncInfo.Key))
            End If
        End SyncLock
        If callbackImage IsNot Nothing Then
            If downloadAsyncInfo.Value IsNot Nothing Then
                downloadAsyncInfo.Value.Invoke(callbackImage)
            End If
            Exit Sub
        End If
        Dim hv As New HttpVarious()
        Dim dlImage As Image = hv.GetImage(downloadAsyncInfo.Key, 10000)
        SyncLock lockObject
            If Me.innerDictionary(downloadAsyncInfo.Key) Is Nothing Then Me.innerDictionary(downloadAsyncInfo.Key) = dlImage
            callbackImage = New Bitmap(dlImage)
        End SyncLock
        If downloadAsyncInfo.Value IsNot Nothing Then
            downloadAsyncInfo.Value.Invoke(callbackImage)
        End If
    End Sub
End Class

File: Hoehoe2/ImageDictionary.cs
// Hoehoe - Client of Twitter
// Copyright (c) 2007-2011 kiri_feather (@kiri_feather) <kiri.feather@gmail.com>
//           (c) 2008-2011 Moz (@syo68k)
//           (c) 2008-2011 takeshik (@takeshik) <http://www.takeshik.org/>
//           (c) 2010-2011 anis774 (@anis774) <http://d.hatena.ne.jp/anis774/>
//           (c) 2010-2011 fantasticswallow (@f_swallow) <http://twitter.com/f_swallow>
//           (c) 2011- t.ashula (@t_ashula) <office@ashula.info>
//
// All rights reserved.
// This file is part of Hoehoe.
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 3 of the License, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program. If not, see <http://www.gnu.org/licenses/>, or write to
// the Free Software Foundation, Inc., 51 Franklin Street - Fifth Floor,
// Boston, MA 02110-1301, USA.

namespace Hoehoe
{
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Collections.Specialized;
    using System.Drawing;
    using System.Runtime.Caching;
    using System.Threading;

    public class ImageDictionary : IDictionary<string, Image>, IDisposable
    {
        private readonly object _lockObject = new object();
        private readonly CacheItemPolicy _cachePolicy = new CacheItemPolicy();
        private readonly MemoryCache _innerDictionary;
        private readonly Semaphore _netSemaphore;
        private bool _pauseGetImage; // 取得一時停止
        private bool _popping;
        private long _removedCount;
        private readonly Stack<KeyValuePair<string, Action<Image>>> waitStack;

        public ImageDictionary(int cacheMemoryLimit)
        {
            lock (_lockObject)
            {
                // 5Mb,80%
                // キャッシュチェック間隔はデフォルト値（2分毎）
                _innerDictionary = new MemoryCache(
                    "imageCache",
                    new NameValueCollection
                    {
                        { "CacheMemoryLimitMegabytes", cacheMemoryLimit.ToString() },
                        { "PhysicalMemoryLimitPercentage", "80" }
                    });
                waitStack = new Stack<KeyValuePair<string, Action<Image>>>();
                _cachePolicy.RemovedCallback = CacheRemoved;
                _cachePolicy.SlidingExpiration = TimeSpan.FromMinutes(30); // 30分参照されなかったら削除
                _netSemaphore = new Semaphore(5, 5);
            }
        }

        public delegate void GetImageDelegate(KeyValuePair<string, Action<Image>> arg1);

        public long CacheCount
        {
            get { return _innerDictionary.GetCount(); }
        }

        public long CacheMemoryLimit
        {
            get { return _innerDictionary.CacheMemoryLimit; }
        }

        public long CacheRemoveCount
        {
            get { return _removedCount; }
        }

        public int Count
        {
            get
            {
                lock (_lockObject)
                {
                    return Convert.ToInt32(_innerDictionary.GetCount());
                }
            }
        }

        public bool IsReadOnly
        {
            get { return false; }
        }

        public ICollection<string> Keys
        {
            get
            {
                throw new NotImplementedException();
            }
        }

        public bool PauseGetImage
        {
            get
            {
                return _pauseGetImage;
            }

            set
            {
                if (!_pauseGetImage && !_popping)
                {
                    // 最新から処理し
                    _popping = true;
                    ThreadStart imgDlProc = () =>
                    {
                        while (!_pauseGetImage)
                        {
                            if (waitStack.Count > 0)
                            {
                                KeyValuePair<string, Action<Image>> req = default(KeyValuePair<string, Action<Image>>);
                                lock (_lockObject)
                                {
                                    req = waitStack.Pop();
                                }

                                if (Configs.Instance.IconSz == IconSizes.IconNone)
                                {
                                    continue;
                                }

                                var proc = new GetImageDelegate(GetImage);
                                try
                                {
                                    _netSemaphore.WaitOne();
                                }
                                catch (Exception)
                                {
                                    // Disposed,Release漏れ
                                }

                                proc.BeginInvoke(req, null, null);
                            }
                            else
                            {
                                Thread.Sleep(100);
                            }
                        }

                        _popping = false;
                    };

                    imgDlProc.BeginInvoke(null, null);
                }
            }
        }

        public long PhysicalMemoryLimit
        {
            get { return _innerDictionary.PhysicalMemoryLimit; }
        }

        public TimeSpan PollingInterval
        {
            get { return _innerDictionary.PollingInterval; }
        }

        public ICollection<Image> Values
        {
            get { throw new NotImplementedException(); }
        }

        public Image this[string key, bool force, Action<Image> callBack]
        {
            get
            {
                lock (_lockObject)
                {
                    if (force)
                    {
                        _innerDictionary.Remove(key);
                    }
                    else
                    {
                        if (_innerDictionary.Contains(key))
                        {
                            return (Image)_innerDictionary[key];
                        }
                    }

                    // スタックに積む
                    waitStack.Push(new KeyValuePair<string, Action<Image>>(key, callBack));
                }

                return null;
            }
        }

        public Image this[string key]
        {
            get
            {
                lock (_lockObject)
                {
                    if (_innerDictionary[key] == null)
                    {
                        return null;
                    }

                    try
                    {
                        return (Image)_innerDictionary[key];
                    }
                    catch (Exception)
                    {
                        _innerDictionary.Remove(key);
                        return null;
                    }
                }
            }

            set
            {
                lock (_lockObject)
                {
                    _innerDictionary.Remove(key);
                    _innerDictionary.Add(key, value, _cachePolicy);
                }
            }
        }

        public void Add(KeyValuePair<string, Image> item)
        {
            Add(item.Key, item.Value);
        }

        public void Add(string key, Image value)
        {
            lock (_lockObject)
            {
                if (_innerDictionary.Contains(key))
                {
                    return;
                }

                _innerDictionary.Add(key, value, _cachePolicy);
            }
        }

        public void Clear()
        {
            lock (_lockObject)
            {
                _innerDictionary.Trim(100);
            }
        }

        public bool Contains(KeyValuePair<string, Image> item)
        {
            lock (_lockObject)
            {
                return _innerDictionary.Contains(item.Key) && ReferenceEquals(_innerDictionary[item.Key], item.Value);
            }
        }

        public bool ContainsKey(string key)
        {
            return _innerDictionary.Contains(key);
        }

        public void CopyTo(KeyValuePair<string, Image>[] array, int arrayIndex)
        {
            lock (_lockObject)
            {
                throw new NotImplementedException();
            }
        }

        public void Dispose()
        {
            lock (_lockObject)
            {
                _netSemaphore.Dispose();
                _innerDictionary.Dispose();
                GC.SuppressFinalize(this);
            }
        }

        public IEnumerator<KeyValuePair<string, Image>> GetEnumerator()
        {
            throw new NotImplementedException();
        }

        public IEnumerator GetEnumerator1()
        {
            throw new NotImplementedException();
        }

        public bool Remove(KeyValuePair<string, Image> item)
        {
            return Remove(item.Key);
        }

        public bool Remove(string key)
        {
            lock (_lockObject)
            {
                _innerDictionary.Remove(key);
            }

            return true;
        }

        public bool TryGetValue(string key, out Image value)
        {
            lock (_lockObject)
            {
                if (_innerDictionary.Contains(key))
                {
                    value = (Image)_innerDictionary[key];
                    return true;
                }

                value = default(Image);
                return false;
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator1();
        }

        private void CacheRemoved(CacheEntryRemovedArguments item)
        {
            ((Image)item.CacheItem.Value).Dispose();
            _removedCount += 1;
        }

        private void GetImage(KeyValuePair<string, Action<Image>> downloadAsyncInfo)
        {
            Image callbackImage = null;
            lock (_lockObject)
            {
                if (_innerDictionary[downloadAsyncInfo.Key] != null)
                {
                    callbackImage = (Image)_innerDictionary[downloadAsyncInfo.Key];
                }
            }

            if (callbackImage != null)
            {
                if (downloadAsyncInfo.Value != null)
                {
                    downloadAsyncInfo.Value.Invoke(callbackImage);
                }

                _netSemaphore.Release();
                return;
            }

            var hv = new HttpVarious();
            Image image = hv.GetImage(downloadAsyncInfo.Key, 10000);
            lock (_lockObject)
            {
                if (_innerDictionary[downloadAsyncInfo.Key] == null)
                {
                    if (image != null)
                    {
                        _innerDictionary.Add(downloadAsyncInfo.Key, image, _cachePolicy);
                        callbackImage = image;
                    }
                }
                else
                {
                    callbackImage = (Image)_innerDictionary[downloadAsyncInfo.Key];
                }
            }

            if (downloadAsyncInfo.Value != null)
            {
                downloadAsyncInfo.Value.Invoke(callbackImage);
            }

            _netSemaphore.Release();
        }
    }
}